source/Plugins.cpp

```diff
@@ -164,3 +186,80 @@ void Plugins::TogglePlugin(const string &name)
 	auto *plugin = plugins.Get(name);
 	plugin->currentState = !plugin->currentState;
 }
+
+
+
+future<void> Plugins::Install(const InstallData &installData, bool guarded)
+{
+	oldNetworkActivity = true;
+	if(!guarded)
+	{
+		lock_guard<mutex> guard(activePluginsMutex);
+		if(activePlugins.count(installData.name))
+			return future<void>();
+		else
+			activePlugins.insert(installData.name);
+
+		// Create a new entry for the plugin.
+		Plugin *newPlugin = plugins.Get(installData.name);
+		newPlugin->name = installData.name;
+		newPlugin->aboutText = installData.aboutText;
+		newPlugin->path = Files::Plugins() + installData.name + "/";
+		newPlugin->version = installData.version;
+		newPlugin->enabled = false;
+		newPlugin->currentState = true;
+	}
+	return async(launch::async, [installData]() noexcept -> void
+		{
+			string zipLocation = Files::Plugins() + installData.name + ".zip";
+			bool success = PluginHelper::Download(installData.url, zipLocation);
+			if(success)
+			{
+				success = PluginHelper::ExtractZIP(
+					zipLocation,
+					Files::Plugins(), installData.name + "/");
+				if(success)
+					Files::Write(Files::Plugins() + installData.name + "/version.txt", installData.version);
+				else
+					Files::DeleteDir(Files::Plugins() + installData.name);
+			}
+			Files::Delete(zipLocation);
+			{
+				lock_guard<mutex> guard(activePluginsMutex);
+				activePlugins.erase(installData.name);
+			}
+		});
+}
+
+
+
+future<void> Plugins::Update(const InstallData &installData)
+{
+	{
+		lock_guard<mutex> guard(activePluginsMutex);
+		if(activePlugins.count(installData.name))
+			return future<void>();
+		else
+			activePlugins.insert(installData.name);
```

Comment:
Insert returns a pair containing a bool whether the element already existed. So you can always insert and then check the return value to know whether to return or not

https://en.cppreference.com/w/cpp/container/set/insert

Reply:
that's handy